api/src/main/java/org/openmrs/Cohort.java

```diff
@@ -180,43 +181,68 @@ public void removeMember(Integer memberId) {
 		removeMember(new Patient(memberId));
 	}
 	
+	/**
+	 * @since 2.1.0
+	 */
 	public void removeMember(Patient patient) {
-		List<CohortMembership> memberToRemoveList = getMembers().stream()
-				.filter(m -> m.getPatient().getPatientId().equals(patient.getPatientId())).collect(Collectors.toList());
-		memberToRemoveList.forEach(m -> m.setEndDate(new Date()));
+		CohortMembership memberToRemove = this.getActiveMembership(patient);
+		memberToRemove.setEndDate(new Date());
 	}
-
+	
+	/**
+	 * @since 2.1.0
+	 */
 	public void addMembership(CohortMembership cohortMembership) {
 		if (!this.contains(cohortMembership.getPatient().getPatientId())) {
 			cohortMembership.setCohort(this);
-			getMembers().add(cohortMembership);
+			getMemberships().add(cohortMembership);
 		}
 	}
-
+	
+	/**
+	 * @since 2.1.0
+	 */
 	public void removeMembership(CohortMembership cohortMembership) {
-		List<CohortMembership> memberToRemoveList = getMembers().stream()
-				.filter(m -> m.equals(cohortMembership) ||
-						m.getPatient().getPatientId().equals(cohortMembership.getPatient().getPatientId()))
-				.collect(Collectors.toList());
-		memberToRemoveList.forEach(m -> m.setEndDate(new Date()));
+		cohortMembership.setEndDate(new Date());
 	}
-
+	
+	/**
+	 * @since 2.1.0
+	 */
 	public List<CohortMembership> getMemberships(Date asOf) {
-		return getMembers().stream()
-				.filter(m -> !m.getStartDate().before(asOf)).collect(Collectors.toList());
+		return getMemberships().stream()
+				.filter(m -> !m.getStartDate().before(asOf) &&
+						m.getStartDate().equals(asOf) &&
+						!m.getVoided() && m.getEndDate() == null)
+				.collect(Collectors.toList());
 	}
-
+	
+	/**
+	 * @since 2.1.0
+	 */
+	public CohortMembership getActiveMembership(Patient patient) {
```

Comment:
I believe I created this to get an active membership for a patient as suggested by you in the earlier commit. Since I was performing this query frequently, should I keep this method around? I have simplified the method to:

```
		return getMemberships().stream()
				.filter(m -> m.isMembershipActive() && m.getPatient().getPatientId().equals(patient.getPatientId()))
				.collect(Collectors.toList())
				.get(0);
```